### ISSUE TITLE
Missing equals and hashcode in MapEnvelope and CollectionEnvelope

### ISSUE THREAD

**Issue by krzyk** [yegor256/cactoos#844]
Following test cases fail:
```
        MatcherAssert.assertThat(
            new CollectionOf("foo"),
            new IsEqual<>(new CollectionOf("foo"))
        );
```
```
        MatcherAssert.assertThat(
            new MapOf(new MapEntry("foo", "bar")),
            new IsEqual<>(new MapEntry("foo", "bar"))
        );
```
But they should work, because collection and map is just a container.

**Comment by 0crat** [comment]
@llorllale[/z](https://www.0crat.com/u/llorllale) please, pay attention to this issue

**Comment by llorllale** [comment]
@krzyk agreed, `CollectionEnvelope` and `MapEnvelope` should implement equals/hashcode

**Comment by llorllale** [comment]
@0crat in

**Comment by 0crat** [comment]
Job [#844](https://github.com/yegor256/cactoos/issues/844) is now in scope, role is `DEV`

**Comment by 0crat** [comment]
> @0crat in ([here](https://github.com/yegor256/cactoos/issues/844#issuecomment-387824794))

@llorllale Job [#844](https://github.com/yegor256/cactoos/issues/844) is already in scope

**Comment by 0crat** [comment]
Bug was reported, see [§29](http://www.zerocracy.com/policy.html#29): +15 point(s) just awarded to @krzyk[/z](https://www.0crat.com/u/krzyk)

**Comment by 0crat** [comment]
The job [#844](https://github.com/yegor256/cactoos/issues/844) assigned to @proshin-roman[/z](https://www.0crat.com/u/proshin-roman), here is [why](https://www.0crat.com/footprint/C63314D6Z/1301946369); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html)

**Comment by 0crat** [comment]
@proshin-roman[/z](https://www.0crat.com/u/proshin-roman) this job was assigned to you 6days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by 0pdd** [comment]
@krzyk the puzzle [#881](https://github.com/yegor256/cactoos/issues/881) is still not solved.

**Comment by proshin-roman** [comment]
@krzyk One part of the issue has been fixed in #865 and a new puzzle has been created for another part. Please, close the ticket.

**Comment by krzyk** [comment]
@proshin-roman thanks

**Comment by 0crat** [comment]
@elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova) please review this job completed by @proshin-roman[/z](https://www.0crat.com/u/proshin-roman), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#844](https://github.com/yegor256/cactoos/issues/844) is now out of scope

**Comment by elenavolokhova** [comment]
@0crat quality good

**Comment by 0pdd** [comment]
@krzyk the puzzle [#904](https://github.com/yegor256/cactoos/issues/904) is still not solved; solved: [#881](https://github.com/yegor256/cactoos/issues/881).

**Comment by 0pdd** [comment]
@krzyk all 2 puzzles are solved here: [#881](https://github.com/yegor256/cactoos/issues/881), [#904](https://github.com/yegor256/cactoos/issues/904).
